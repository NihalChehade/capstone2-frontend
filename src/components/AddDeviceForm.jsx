import React, { useState, useContext } from "react";
import {
  Form,
  FormGroup,
  Label,
  Input,
  Button,
  FormFeedback,
  Spinner,
  Alert
} from "reactstrap";
import UserContext from "../UserContext";
import { useNavigate } from "react-router-dom";
import "./AddDeviceForm.css";

const AddDeviceForm = () => {
  const { addDevice } = useContext(UserContext);
  const [formData, setFormData] = useState({
    name: "",
    serial_number: "",
    type: "",
    room: "",
    status: "",
  });

  const navigate = useNavigate();

  const [formError, setFormError] = useState({});
  const [loading, setLoading] = useState(false);

  // Handle input changes
  const handleChange = (e) => {
    const { name, value } = e.target;
    setFormData((data) => ({
      ...data,
      [name]: value,
    }));
    // Clear the specific field error when user changes the input
    if (formError[name]) {
      setFormError({ ...formError, [name]: undefined });
    }
  };

  // Handle form submission
  const handleSubmit = async (e) => {
    e.preventDefault();
    setLoading(true); // Start spinner
    try {
      const success = await addDevice(formData);
      if (success) {
        navigate("/"); // Redirect after successful device addition
      }
    } catch (error) {
      setFormError(error); // Set form errors based on the error object received
    } finally {
      setLoading(false); // Stop spinner regardless of the outcome
    }
  };

  return (
    <div className="AddDeviceForm">
      <h2>Add Device Form</h2>
      {loading && <div className="text-center"><Spinner color="primary" /></div>}
      {/* Display a general error message if there is one */}
       {formError.general && <Alert color="danger" >{formError.general}</Alert>}
      <Form onSubmit={handleSubmit}>
        {Object.entries(formData).map(([key, value]) => (
          <FormGroup key={key}>
            <Label for={key}>{key.replace('_', ' ').replace(/\b\w/g, l => l.toUpperCase())}</Label>
            <Input
              type="text"
              bsSize="lg"
              name={key}
              id={key}
              value={value}
              onChange={handleChange}
              invalid={!!formError[key]}
              required
            />
            <FormFeedback>{formError[key]}</FormFeedback>
          </FormGroup>
        ))}

        <Button
          type="submit"
          size="lg"
          color="primary"
          block
          disabled={loading}
        >
          ADD
        </Button>
      </Form>
    </div>
  );
};

export default AddDeviceForm;
