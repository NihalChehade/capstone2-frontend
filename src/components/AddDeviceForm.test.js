import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";
import UserContext from "../UserContext";
import AddDeviceForm from "./AddDeviceForm";
import { BrowserRouter } from "react-router-dom";

// Mock UserContext
const mockAddDevice = jest.fn();
const wrapper = ({ children }) => (
  <BrowserRouter>
    <UserContext.Provider value={{ addDevice: mockAddDevice }}>
      {children}
    </UserContext.Provider>
  </BrowserRouter>
);

describe("AddDeviceForm", () => {
  it("renders without crashing and matches snapshot", () => {
    const { asFragment } = render(<AddDeviceForm />, { wrapper });
    expect(asFragment()).toMatchSnapshot();
  });

  it("submits data correctly", async () => {
    const mockAddDevice = jest.fn().mockResolvedValue(true);
    const { getByLabelText, getByRole } = render(
      <BrowserRouter>
      <UserContext.Provider value={{ addDevice: mockAddDevice }}>
        <AddDeviceForm />
      </UserContext.Provider>
      </BrowserRouter>
    );

    fireEvent.change(getByLabelText(/Name/i), {
      target: { value: "New Light" },
    });
    fireEvent.change(getByLabelText(/Serial Number/i), {
      target: { value: "123456789" },
    });
    fireEvent.change(getByLabelText(/Type/i), { target: { value: "Light" } });
    fireEvent.change(getByLabelText(/Room/i), {
      target: { value: "Living Room" },
    });
    fireEvent.change(getByLabelText(/Status/i), { target: { value: "on" } });

    fireEvent.click(getByRole("button", { name: /add/i }));

    await waitFor(() =>
      expect(mockAddDevice).toHaveBeenCalledWith({
        name: "New Light",
        serial_number: "123456789",
        type: "Light",
        room: "Living Room",
        status: "on",
      })
    );
  });
});
