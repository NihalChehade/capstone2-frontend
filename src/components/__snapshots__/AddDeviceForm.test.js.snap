// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddDeviceForm renders without crashing and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="AddDeviceForm"
  >
    <h2>
      Add Device Form
    </h2>
    <form
      class=""
    >
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="name"
        >
          Name
        </label>
        <input
          aria-invalid="false"
          class="form-control-lg form-control"
          id="name"
          name="name"
          required=""
          type="text"
          value=""
        />
        <div
          class="invalid-feedback"
        />
      </div>
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="serial_number"
        >
          Serial Number
        </label>
        <input
          aria-invalid="false"
          class="form-control-lg form-control"
          id="serial_number"
          name="serial_number"
          required=""
          type="text"
          value=""
        />
        <div
          class="invalid-feedback"
        />
      </div>
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="type"
        >
          Type
        </label>
        <input
          aria-invalid="false"
          class="form-control-lg form-control"
          id="type"
          name="type"
          required=""
          type="text"
          value=""
        />
        <div
          class="invalid-feedback"
        />
      </div>
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="room"
        >
          Room
        </label>
        <input
          aria-invalid="false"
          class="form-control-lg form-control"
          id="room"
          name="room"
          required=""
          type="text"
          value=""
        />
        <div
          class="invalid-feedback"
        />
      </div>
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="status"
        >
          Status
        </label>
        <input
          aria-invalid="false"
          class="form-control-lg form-control"
          id="status"
          name="status"
          required=""
          type="text"
          value=""
        />
        <div
          class="invalid-feedback"
        />
      </div>
      <button
        class="btn btn-primary btn-lg d-block w-100"
        type="submit"
      >
        ADD
      </button>
    </form>
  </div>
</DocumentFragment>
`;
